src/php/parser/Parser.php

```diff
@@ -0,0 +1,530 @@
+        }
+
+        return $this->match();
+    }
+
+    private function match(): AST
+    {
+        $action = $this->moveEval($this->evalNext($this->automata[$this->start]->getExpression(), 0, new ASTs(), new RawError(0, array(), new MatchErrors(), $this->start), new Continuations()));
+
+        while (true) {
+            switch ($action->getType()) {
+                case ActionType::EVAL:
+                {
+                    $action = $this->moveEval(EvalAction::asEvalAction($action));
+                    break;
+                }
+                case ActionType::APPLY:
+                {
+                    $action = ApplyAction::asApplyAction($action);
+                    $continuations = $action->getContinuations();
+                    $matchResult = $action->getMatchResult();
+
+                    if ($continuations->isEmpty()) {
+                        return $this->moveReturn($matchResult);
+                    }
+
+                    $action = $this->moveApply($matchResult, $continuations->head(), $continuations->tail());
+                    break;
+                }
+                default:
+                {
+                    $type = $action->getType();
+                    throw new RuntimeException("Unsupported action type $type!");
+                }
+            }
+        }
+
+        throw new RuntimeException("must not be here!");
+    }
+
+    private function moveEval(EvalAction $action): Action
+    {
+        $expression = $action->getExpression();
+        $position = $action->getPosition();
+        $asts = $action->getAsts();
+        $error = $action->getError();
+        $continuations = $action->getContinuations();
+        $eof = $position >= strlen($this->input);
+
+        switch ($expression->getType()) {
+            case ExpressionType::ANY_CHAR:
+            {
+                if ($eof) {
+                    $matchResult = $this->reject($this->updateError($error, $position, new WildcardError()));
+                } else {
+                    $expected = mb_substr($this->input, $position, 1);
+                    $ord = IntlChar::ord($expected);
+
+                    $matchResult = $this->accept($position + strlen($expected), ASTs::asts(new Char($ord, $position), $asts), $error);
+                }
+
+                return $this->applyNext($continuations, $matchResult);
+            }
+            case ExpressionType::ALT:
+            {
+                $expressions = Expression::asAltExpression($expression)->getExpressions();
+
+                if ($expressions->isEmpty()) {
+                    return $this->applyNext($continuations, $this->reject($error));
+                } else {
+                    $cons = Continuations::cons(new AltContinuation($expressions->tail(), $position, $asts), $continuations);
+                    return $this->evalNext($expressions->head(), $position, $asts, $error, $cons);
+                }
+            }
+            case ExpressionType:: AND:
```

Comment:
Redundant space between `::` and `AND`

Reply:
resolved in d9e587ee5bc4b9568109af96eb9eb692c762e509